src/Microsoft.AspNetCore.OData/Common/TypeHelper.cs

```diff
@@ -40,7 +41,136 @@ public static bool IsDynamicTypeWrapper(this Type type)
+                        && genericTypeDefinition.ImplementsInterface(typeof(IGroupByWrapper<,>)),
+                        "ComputeWrapper<T> must inherit from DynamicTypeWrapper and implement IComputeWrapper<T> and IGroupByWrapper<TContainer, TWrapper>");
+
+                    return true;
+                }
+
+                // Custom implementation
+                // Must inherit from DynamicTypeWrapper
+                // Must implement IComputeWrapper<T> and IGroupByWrapper<TContainer, TWrapper>
+                if (typeof(DynamicTypeWrapper).IsAssignableFrom(genericTypeDefinition) &
+                    genericTypeDefinition.ImplementsInterface(typeof(IComputeWrapper<>))
+                    && genericTypeDefinition.ImplementsInterface(typeof(IGroupByWrapper<,>)))
+                {
+                    entityType = typeToCheck.GetGenericArguments()[0];
+
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Determines whether the specified type is a <see cref="FlatteningWrapper{T}"/> or a custom implementation
+        /// that inherits from <see cref="DynamicTypeWrapper"/> and implements both <see cref="IFlatteningWrapper{T}"/>
+        /// and <see cref="IGroupByWrapper{TContainer, TWrapper}"/>.
+        /// </summary>
+        /// <param name="typeToCheck">The type to check.</param>
+        /// <returns>
+        /// <c>true</c> if the specified type is a <see cref="FlatteningWrapper{T}"/> or a custom implementation
+        /// that meets the criteria; otherwise, <c>false</c>.
+        /// </returns>
+        public static bool IsFlatteningWrapper(this Type typeToCheck)
+        {
+            if (typeToCheck == null)
+            {
+                return false;
+            }
+
+            if (typeToCheck.IsGenericType)
+            {
+                Type genericTypeDefinition = typeToCheck.GetGenericTypeDefinition();
+
+                // Default implementation
+                if (genericTypeDefinition == typeof(FlatteningWrapper<>))
+                {
+                    Debug.Assert(
+                        typeof(DynamicTypeWrapper).IsAssignableFrom(genericTypeDefinition)
+                        && genericTypeDefinition.ImplementsInterface(typeof(IFlatteningWrapper<>))
+                        && genericTypeDefinition.ImplementsInterface(typeof(IGroupByWrapper<,>)),
+                        "FlatteningWrapper<T> must inherit from DynamicTypeWrapper and implement IFlatteningWrapper<T> and IGroupByWrapper<TContainer, TWrapper>");
+
+                    return true;
+                }
+
+                // Custom implementation
+                // Must inherit from DynamicTypeWrapper
+                // Must implement IFlatteningWrapper<T> and IGroupByWrapper<TContainer, TWrapper>
+                return typeof(DynamicTypeWrapper).IsAssignableFrom(genericTypeDefinition) &
```

Comment:
```suggestion
                return typeof(DynamicTypeWrapper).IsAssignableFrom(genericTypeDefinition) &8 
```
Will this avoid the second statement getting executed if the first one fails.

Reply:
The `&` was inadvertent. I intended `&&`. Good catch!